server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -412,11 +417,68 @@ public void executeQueryPhase(ShardSearchRequest request, SearchShardTask task,
                     return;
                 }
             }
-            // fork the execution in the search thread pool
-            runAsync(getExecutor(shard), () -> executeQueryPhase(orig, task), l);
+            ensureAfterSeqNoRefreshed(shard, orig, () -> executeQueryPhase(orig, task), l);
         }));
     }
 
+    private <T> void ensureAfterSeqNoRefreshed(IndexShard shard, ShardSearchRequest request, CheckedSupplier<T, Exception> executable,
+                                               ActionListener<T> listener) {
+        final Executor executor = getExecutor(shard);
+        executor.execute(new ActionRunnable<>(listener) {
+            @Override
+            protected void doRun() {
+                if (request.waitForCheckpoint() > SequenceNumbers.NO_OPS_PERFORMED) {
+                    final AtomicBoolean isDone = new AtomicBoolean(false);
+                    final AtomicReference<Scheduler.ScheduledCancellable> timeoutTask = new AtomicReference<>();
+                    final ActionListener<Void> readyListener = new ActionListener<>() {
+
+                        @Override
+                        public void onResponse(Void unused) {
+                            if (isDone.compareAndSet(false, true)) {
+                                Scheduler.ScheduledCancellable localTimeoutTask = timeoutTask.get();
+                                if (localTimeoutTask != null) {
+                                    localTimeoutTask.cancel();
+                                }
+                                runAsync(executor, executable, listener);
+                            }
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            if (isDone.compareAndSet(false, true)) {
+                                Scheduler.ScheduledCancellable localTimeoutTask = timeoutTask.get();
+                                if (localTimeoutTask != null && e instanceof ElasticsearchTimeoutException == false) {
+                                    localTimeoutTask.cancel();
+                                }
+                                listener.onFailure(e);
+                            }
+                        }
+                    };
+                    shard.addRefreshListener(request.waitForCheckpoint(), readyListener);
+                    SearchSourceBuilder source = request.source();
+                    final TimeValue timeout;
+                    if (source == null) {
+                        timeout = defaultSearchTimeout;
+                    } else {
+                        timeout = source.timeout() == null ? defaultSearchTimeout : source.timeout();
```

Comment:
we shouldn't reuse the timeout option here. It has a special meaning that doesn't apply here. Let's add a new parameter, something like `wait_for_checkpoint_timeout` ?